### PR TITLE
Various essential fixes

### PR DESCRIPTION
**Pull request by dgreene-r7**
This PR adds a new method for secret injection that properly handles arbitrarily nested properties and arrays. Using this method is gated behind an explicit configuration option so consumers can opt in if they need that functionality.

It also contains a bunch of minor fixes:
- Convert to go modules, deprecate dep.
- Add some quality-of-life improvements around configuring log level
- Add recovery and request logging middleware
- Fix up error handling so we don't swallow certain errors we should at least log
- Improved make targets, added linting, benchtesting, testing with coverage, etc.
- Adds `HEAD` request handlers to every route.
- Adds support for KMS encrypted properties.

The `HEAD` request handlers solves an interesting problem we noticed where wget's `--spider` has inconsistent behaviors across versions. With the version of wget that ships with Alpine (which is actually the BusyBox version), the `--spider` flag issues a `GET` request and simply swallows the output. However, GNU wget issues a `HEAD` request. This isn't really a big deal but we've seen wget used for health checks so in order to ensure consistent behavior, we should respond to both.

### CODE REVIEW

File: watchers/v2/s3/s3.go [hunk 1]
```diff
@@ -186,27 +211,186 @@ func getPropertyFiles(files []string, b string, svc s3iface.S3API, log *zap.Logg
...
+					continue
+				}
+				if _, ok := s[secret.KMSIdentifier]; ok {
+					decrypted, err := handleKMSSecret(ctx, s)
+					if err != nil {
+						log.Error("error handling KMS secret",
+							zap.Error(err),
+							zap.String("key", k),
+						)
+						// TODO: Should we log the error and continue, with a key that doesn't have a value or
```
Comment (lmilby-r7) [inline watchers/v2/s3/s3.go:327]: would InjectSecretsV2 handle this?

File: api/v2/health/health.go [hunk 2]
```diff
@@ -22,21 +22,33 @@ func GetHealthz(w http.ResponseWriter, r *http.Request, log *zap.Logger) {
...
 	if err != nil {
 		log.Error("Failed to unmarshal json",
 			zap.Error(err),
 		)
+		w.WriteHeader(http.StatusInternalServerError)
+		if r.Method == http.MethodHead {
+			return
+		}
+
+		w.Write([]byte(`{}`)) //nolint: errcheck
```
Comment (cmchugh-r7) [inline api/v2/health/health.go:40]: why is this needed?

File: main.go [hunk 3]
```diff
@@ -1,42 +1,75 @@
...
+	"fmt"
 	"net/http"
 
+	"github.com/gorilla/handlers"
 	"github.com/gorilla/mux"
 	"github.com/spf13/viper"
 	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/rapid7/cps/api"
```
Comment (cmchugh-r7) [inline main.go:14]: import needs grouping

File: main.go [hunk 4]
```diff
@@ -1,42 +1,75 @@
...
+	flag.StringVar(&configFile, "config", "", "(Optional) Config file")
+	flag.StringVar(&configFile, "c", "", "(Optional) Config file")
+	flag.Parse()
 
 	viper.SetConfigName("cps")
 	viper.AddConfigPath("/etc/cps/")
 	viper.AddConfigPath(".")
+	viper.SetEnvPrefix("cps")
+	// Allow dev mode to be set via env var
+	viper.BindEnv("dev") //nolint: errcheck
```
Comment (cmchugh-r7) [inline main.go:41]: it seems like this would always set it to dev mode, am i mistaken?

**Review by cmchugh-r7 — APPROVED** [review]
couple questions but seems good

File: api/v2/health/health.go (hunk 2 above)
Comment (dgreene-r7) [reply api/v2/health/health.go:40]: The goal was to ensure a consistent json response even if it is just an empty object. I don't think it's necessary so I can definitely remove it if you think that makes sense.

Comment (cmchugh-r7) [reply api/v2/health/health.go:40]: i don't have strong preferences, just wondering if there was a specific reason 👍  thanks!

File: main.go (hunk 4 above)
Comment (dgreene-r7) [reply main.go:41]: `BindEnv` binds the value of the `dev` key to an env var reader. What this means, in our case, is that if the environment variable `CPS_DEV` (note the call to `SetEnvPrefix`) is set to any value that `strconv.ParseBool` returns as `true`, then we'll be in dev mode.

File: watchers/v2/s3/s3.go (hunk 1 above)
Comment (dgreene-r7) [reply watchers/v2/s3/s3.go:327]: I think it's more of a question as to whether we want to return with a partial set of decrypted secrets and an error or if we should error, set an empty value for that key, and continue. The behavior that's in here now isn't consistent with the behavior in v1 but is, iirc, consistent with the behavior in propsd.

It's more of a question for @asebastian-r7 when he gets back.

**Review by bcall-r7 — APPROVED** [review]
